### ISSUE TITLE
Lowest Level HLO Graph

### ISSUE THREAD

**Issue by codeislife99** [pytorch/xla#2997]
I wanted to get the lowest level HLO graph, which is the graph after all fusions and optimizations have been done. The HLO Dump that I get with the environment variables in the troubleshooting README has aten operations in it. However when I profile the code, I see a bunch of `fusion_xx`(fusion_13, fusion_967 etc) kernels. I would like to know which operators are contributing to these fusion kernels to debug performance issues in XLA. 

In TF, there is an easy way to get this dump, but couldn't find an easy way for PT yet. Thank you.

**Comment by JackCaoG** [comment]
If you set the `XLA_SAVE_TENSORS_FMT ` to `hlo` and set `XLA_SAVE_TENSORS_FILE`, you should be able to get the hlo dump (without any fushion through). Regarding the fused kernel, @jysohn23 I remembered there were some frictions for us to know the op that was fused. Is that still true?

**Comment by codeislife99** [comment]
Hi Jack, thanks for your response. 

I am already doing what you said, but because I am trying to profile bottlenecks, and kernels which are launched are named like `fusion_xx` the graph dumps by enabling those environment variables are not very helpful. This is because I want to correlate lets say `fusion_33` with some real ops (like fc + softmax etc.). The lowest level HLO pass should give me this information like it already does in Tensorflow, but I don't know how I can get that.

**Comment by jysohn23** [comment]
You mention you're profiling. Are you profiling the execution of the graphs on TPU? If you aren't already, can you try running with `XLA_HLO_DEBUG=1`? With that turned on the execution traces should also show additional op metadata on the execution traces in addition to just `fusion_xx`. Ex: https://cloud.google.com/tpu/docs/images/pytorchxlaperf--pltdmhvn17.png

**Comment by codeislife99** [comment]
I am on a GPU. I don't see the correspondence in my trace viewer. `XLA_HLO_DEBUG=1` and `PT_XLA_DEBUG=1` both are already turned on. 
![image](https://user-images.githubusercontent.com/17168743/122117599-e812a500-cddb-11eb-952f-6b603e2dad32.png)

**Comment by codeislife99** [comment]
So it turns out that the way to get the HLO graph before and after each optimization is to simply use the following env variables. 
```
export TF_CPP_MIN_LOG_LEVEL=0
export TF_CPP_VMODULE=hlo_pass_pipeline=1
export XLA_FLAGS="--xla_dump_to=/pytorch/xla/test/hlo_graph --xla_dump_hlo_as_text \
--xla_dump_hlo_pass_re=.*"
```
More options can be found [here](https://docs.rs/tensorflow_proto/0.2.0/tensorflow_proto/xla/struct.DebugOptions.html#)

**Comment by codeislife99** [comment]
This is how a small section of the output will look like. 
```
%fused_computation.14 (param_0.39: f32[128,10,24,24], param_1.52: f32[10]) -> f32[128,10,12,12] {
  %param_0.39 = f32[128,10,24,24]{3,2,1,0} parameter(0)
  %param_1.52 = f32[10]{0} parameter(1)
  %broadcast.55 = f32[128,24,24,10]{2,1,3,0} broadcast(f32[10]{0} %param_1.52), dimensions={3}, metadata={op_type="aten__convolution_overrideable" op_name="train_loop.1/build_graph.1/conv1.1/aten__convolution_overrideable"}
  %transpose.1 = f32[128,10,24,24]{3,2,1,0} transpose(f32[128,24,24,10]{2,1,3,0} %broadcast.55), dimensions={0,3,1,2}, metadata={op_type="aten__convolution_overrideable" op_name="train_loop.1/build_graph.1/conv1.1/aten__convolution_overrideable"}
  %add.13 = f32[128,10,24,24]{3,2,1,0} add(f32[128,10,24,24]{3,2,1,0} %param_0.39, f32[128,10,24,24]{3,2,1,0} %transpose.1), metadata={op_type="aten__convolution_overrideable" op_name="train_loop.1/build_graph.1/conv1.1/aten__convolution_overrideable"}
  %constant_44 = f32[] constant(-inf), metadata={op_type="aten__max_pool2d" op_name="train_loop.1/build_graph.1/conv1.1/aten__max_pool2d"}
  %reduce-window.1 = f32[128,10,12,12]{3,2,1,0} reduce-window(f32[128,10,24,24]{3,2,1,0} %add.13, f32[] %constant_44), window={size=1x1x2x2 stride=1x1x2x2}, to_apply=%max_F32.28, metadata={op_type="aten__max_pool2d" op_name="train_loop.1/build_graph.1/conv1.1/aten__max_pool2d"}
  %constant_45 = f32[] constant(0), metadata={op_type="aten__relu" op_name="train_loop.1/build_graph.1/conv1.1/aten__relu"}
  %broadcast.54 = f32[128,10,12,12]{3,2,1,0} broadcast(f32[] %constant_45), dimensions={}, metadata={op_type="aten__relu" op_name="train_loop.1/build_graph.1/conv1.1/aten__relu"}
  ROOT %maximum.2 = f32[128,10,12,12]{3,2,1,0} maximum(f32[128,10,12,12]{3,2,1,0} %reduce-window.1, f32[128,10,12,12]{3,2,1,0} %broadcast.54), metadata={op_type="aten__relu" op_name="train_loop.1/build_graph.1/conv1.1/aten__relu"}
}

ENTRY %SyncTensorsGraph.191 (p0.1: s64[128], p1.2: f32[10], p2.3: f32[10,50], p3.5: f32[50], p4.6: f32[50,320], p5.8: f32[20], p6.9: f32[20], p7.10: f32[20], p8.11: f32[20], p9.12: f32[20], p10.13: f32[20,10,5,5], p11.14: f32[10], p12.15: f32[10], p13.16: f32[10], p14.17: f32[10], p15.18: f32[10], p16.19: f32[10,1,5,5], p17.20: f32[128,1,28,28]) -> (f32[]) {
  %p5.8 = f32[20]{0} parameter(5), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv2.2/xla__device_data"}
  %p6.9 = f32[20]{0} parameter(6), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv2.2/xla__device_data"}
  %p11.14 = f32[10]{0} parameter(11), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv1.1/xla__device_data"}
  %p12.15 = f32[10]{0} parameter(12), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv1.1/xla__device_data"}
  %p7.10 = f32[20]{0} parameter(7), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv2.2/xla__device_data"}
  %p8.11 = f32[20]{0} parameter(8), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv2.2/xla__device_data"}
  %p13.16 = f32[10]{0} parameter(13), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv1.1/xla__device_data"}
  %p14.17 = f32[10]{0} parameter(14), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv1.1/xla__device_data"}
  %p17.20 = f32[128,1,28,28]{3,2,1,0} parameter(17), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv1.1/xla__device_data"}
  %p16.19 = f32[10,1,5,5]{3,2,1,0} parameter(16), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv1.1/xla__device_data"}
  %custom-call.6 = (f32[128,10,24,24]{3,2,1,0}, u8[0]{0}) custom-call(f32[128,1,28,28]{3,2,1,0} %p17.20, f32[10,1,5,5]{3,2,1,0} %p16.19), window={size=5x5}, dim_labels=bf01_oi01->bf01, custom_call_target="__cudnn$convForward", metadata={op_type="aten__convolution_overrideable" op_name="train_loop.1/build_graph.1/conv1.1/aten__convolution_overrideable"}, backend_config="{\"algorithm\":\"0\",\"tensor_ops_enabled\":false,\"conv_result_scale\":1,\"activation_mode\":\"0\",\"side_input_scale\":0}"
  %get-tuple-element.2 = f32[128,10,24,24]{3,2,1,0} get-tuple-element((f32[128,10,24,24]{3,2,1,0}, u8[0]{0}) %custom-call.6), index=0, metadata={op_type="aten__convolution_overrideable" op_name="train_loop.1/build_graph.1/conv1.1/aten__convolution_overrideable"}
  %p15.18 = f32[10]{0} parameter(15), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv1.1/xla__device_data"}
  %fusion.14 = f32[128,10,12,12]{3,2,1,0} fusion(f32[128,10,24,24]{3,2,1,0} %get-tuple-element.2, f32[10]{0} %p15.18), kind=kLoop, calls=%fused_computation.14, metadata={op_type="aten__relu" op_name="train_loop.1/build_graph.1/conv1.1/aten__relu"}
  %fusion.13 = (f32[10]{0}, f32[10]{0}) fusion(f32[128,10,12,12]{3,2,1,0} %fusion.14), kind=kInput, calls=%fused_computation.13, metadata={op_type="aten__native_batch_norm" op_name="train_loop.1/build_graph.1/conv1.1/aten__native_batch_norm"}
  %get-tuple-element.9 = f32[10]{0} get-tuple-element((f32[10]{0}, f32[10]{0}) %fusion.13), index=1
  %get-tuple-element.8 = f32[10]{0} get-tuple-element((f32[10]{0}, f32[10]{0}) %fusion.13), index=0
  %fusion.11 = f32[128,10,12,12]{3,2,1,0} fusion(f32[10]{0} %p13.16, f32[10]{0} %p14.17, f32[10]{0} %get-tuple-element.9, f32[128,10,12,12]{3,2,1,0} %fusion.14, f32[10]{0} %get-tuple-element.8), kind=kLoop, calls=%fused_computation.11, metadata={op_type="aten__native_batch_norm" op_name="train_loop.1/build_graph.1/conv1.1/aten__native_batch_norm"}
  %p10.13 = f32[20,10,5,5]{3,2,1,0} parameter(10), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv2.2/xla__device_data"}
  %custom-call.7 = (f32[128,20,8,8]{3,2,1,0}, u8[0]{0}) custom-call(f32[128,10,12,12]{3,2,1,0} %fusion.11, f32[20,10,5,5]{3,2,1,0} %p10.13), window={size=5x5}, dim_labels=bf01_oi01->bf01, custom_call_target="__cudnn$convForward", metadata={op_type="aten__convolution_overrideable" op_name="train_loop.1/build_graph.1/conv2.2/aten__convolution_overrideable"}, backend_config="{\"algorithm\":\"1\",\"tensor_ops_enabled\":false,\"conv_result_scale\":1,\"activation_mode\":\"0\",\"side_input_scale\":0}"
  %get-tuple-element.3 = f32[128,20,8,8]{3,2,1,0} get-tuple-element((f32[128,20,8,8]{3,2,1,0}, u8[0]{0}) %custom-call.7), index=0, metadata={op_type="aten__convolution_overrideable" op_name="train_loop.1/build_graph.1/conv2.2/aten__convolution_overrideable"}
  %p9.12 = f32[20]{0} parameter(9), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/conv2.2/xla__device_data"}
  %fusion.10 = f32[128,20,4,4]{3,2,1,0} fusion(f32[128,20,8,8]{3,2,1,0} %get-tuple-element.3, f32[20]{0} %p9.12), kind=kLoop, calls=%fused_computation.10, metadata={op_type="aten__relu" op_name="train_loop.1/build_graph.1/conv2.2/aten__relu"}
  %fusion.9 = (f32[20,16]{1,0}, f32[20,16]{1,0}) fusion(f32[128,20,4,4]{3,2,1,0} %fusion.10), kind=kInput, calls=%fused_computation.9, metadata={op_type="aten__native_batch_norm" op_name="train_loop.1/build_graph.1/conv2.2/aten__native_batch_norm"}
  %get-tuple-element.7 = f32[20,16]{1,0} get-tuple-element((f32[20,16]{1,0}, f32[20,16]{1,0}) %fusion.9), index=1
  %constant_57 = f32[] constant(0), metadata={op_type="aten__relu" op_name="train_loop.1/build_graph.1/conv1.1/aten__relu"}
  %reduce.13 = f32[20]{0} reduce(f32[20,16]{1,0} %get-tuple-element.7, f32[] %constant_57), dimensions={1}, to_apply=%scalar_add_computation.1, metadata={op_type="aten__native_batch_norm" op_name="train_loop.1/build_graph.1/conv2.2/aten__native_batch_norm"}
  %get-tuple-element.6 = f32[20,16]{1,0} get-tuple-element((f32[20,16]{1,0}, f32[20,16]{1,0}) %fusion.9), index=0
  %reduce.11 = f32[20]{0} reduce(f32[20,16]{1,0} %get-tuple-element.6, f32[] %constant_57), dimensions={1}, to_apply=%scalar_add_computation.1, metadata={op_type="aten__native_batch_norm" op_name="train_loop.1/build_graph.1/conv2.2/aten__native_batch_norm"}
  %fusion.7 = f32[128,320]{1,0} fusion(f32[20]{0} %p7.10, f32[20]{0} %p8.11, f32[20]{0} %reduce.13, f32[128,20,4,4]{3,2,1,0} %fusion.10, f32[20]{0} %reduce.11), kind=kLoop, calls=%fused_computation.7, metadata={op_type="aten__view" op_name="train_loop.1/build_graph.1/dense.3/aten__view"}
  %p4.6 = f32[50,320]{1,0} parameter(4), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/dense.3/xla__device_data"}
  %p3.5 = f32[50]{0} parameter(3), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/dense.3/xla__device_data"}
  %broadcast.28 = f32[128,50]{1,0} broadcast(f32[50]{0} %p3.5), dimensions={1}, metadata={op_type="aten__addmm" op_name="train_loop.1/build_graph.1/dense.3/aten__addmm"}
  %custom-call.3 = f32[128,50]{1,0} custom-call(f32[128,320]{1,0} %fusion.7, f32[50,320]{1,0} %p4.6, f32[128,50]{1,0} %broadcast.28), custom_call_target="__cublas$gemm", metadata={op_type="aten__addmm" op_name="train_loop.1/build_graph.1/dense.3/aten__addmm"}, backend_config="{\"alpha_real\":1,\"alpha_imag\":0,\"beta\":1,\"dot_dimension_numbers\":{\"lhs_contracting_dimensions\":[\"1\"],\"rhs_contracting_dimensions\":[\"1\"],\"lhs_batch_dimensions\":[],\"rhs_batch_dimensions\":[]},\"batch_size\":\"1\",\"selected_algorithm\":\"18\"}"
  %fusion.6 = f32[128,50]{1,0} fusion(f32[128,50]{1,0} %custom-call.3), kind=kLoop, calls=%fused_computation.6, metadata={op_type="aten__relu" op_name="train_loop.1/build_graph.1/dense.3/aten__relu"}
  %p2.3 = f32[10,50]{1,0} parameter(2), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/dense.3/xla__device_data"}
  %p1.2 = f32[10]{0} parameter(1), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/dense.3/xla__device_data"}
  %broadcast.29 = f32[128,10]{1,0} broadcast(f32[10]{0} %p1.2), dimensions={1}, metadata={op_type="aten__addmm" op_name="train_loop.1/build_graph.1/dense.3/aten__addmm"}
  %custom-call.5 = f32[128,10]{1,0} custom-call(f32[128,50]{1,0} %fusion.6, f32[10,50]{1,0} %p2.3, f32[128,10]{1,0} %broadcast.29), custom_call_target="__cublas$gemm", metadata={op_type="aten__addmm" op_name="train_loop.1/build_graph.1/dense.3/aten__addmm"}, backend_config="{\"alpha_real\":1,\"alpha_imag\":0,\"beta\":1,\"dot_dimension_numbers\":{\"lhs_contracting_dimensions\":[\"1\"],\"rhs_contracting_dimensions\":[\"1\"],\"lhs_batch_dimensions\":[],\"rhs_batch_dimensions\":[]},\"batch_size\":\"1\",\"selected_algorithm\":\"18\"}"
  %constant_27 = f32[] constant(-inf), metadata={op_type="aten__max_pool2d" op_name="train_loop.1/build_graph.1/conv1.1/aten__max_pool2d"}
  %reduce.136 = f32[128]{0} reduce(f32[128,10]{1,0} %custom-call.5, f32[] %constant_27), dimensions={1}, to_apply=%MaxComputation.132, metadata={op_type="aten__log_softmax" op_name="train_loop.1/build_graph.1/aten__log_softmax"}
  %fusion.4 = f32[128]{0} fusion(f32[128,10]{1,0} %custom-call.5, f32[128]{0} %reduce.136), kind=kLoop, calls=%fused_computation.4, metadata={op_type="aten__log_softmax" op_name="train_loop.1/build_graph.1/aten__log_softmax"}
  %p0.1 = s64[128]{0} parameter(0), metadata={op_type="xla__device_data" op_name="train_loop.1/build_graph.1/xla__device_data"}
  %fusion.2 = (f32[], f32[]) fusion(f32[128]{0} %fusion.4, s64[128]{0} %p0.1, f32[128,10]{1,0} %custom-call.5, f32[128]{0} %reduce.136), kind=kInput, calls=%fused_computation.2, metadata={op_type="aten__nll_loss" op_name="train_loop.1/build_graph.1/aten__nll_loss"}
  %get-tuple-element.4 = f32[] get-tuple-element((f32[], f32[]) %fusion.2), index=0
  %get-tuple-element.5 = f32[] get-tuple-element((f32[], f32[]) %fusion.2), index=1
  %fusion = f32[] fusion(f32[] %get-tuple-element.4, f32[] %get-tuple-element.5), kind=kLoop, calls=%fused_computation, metadata={op_type="aten__nll_loss" op_name="train_loop.1/build_graph.1/aten__nll_loss"}
  ROOT %tuple.190 = (f32[]) tuple(f32[] %fusion)
}
```

**Comment by jysohn23** [comment]
Thanks for posting that info on this issue.
